// backendless/lib/bl/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _desc, _value, _obj, _init, _dec2, _desc2, _value2, _obj2, _init2;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _constants = require('./constants');

var _dispatchEvent = require('./dispatch-event');

var _invokeServiceMethod = require('./invoke-service-method');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var BL = {
  ExecutionTypes: _constants.ExecutionTypes,

  CustomServices: (_dec = (0, _decorators.deprecated)('Backendless.CustomServices', 'Backendless.CustomServices.invoke'), (_obj = {
    invokeSync: _utils2.default.synchronized(_invokeServiceMethod.invokeServiceMethod),
    invoke: _utils2.default.promisified(_invokeServiceMethod.invokeServiceMethod)

  }, (_applyDecoratedDescriptor(_obj, 'invokeSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'invokeSync'), _init = _init ? _init.value : undefined, {
    enumerable: true,
    configurable: true,
    writable: true,
    initializer: function initializer() {
      return _init;
    }
  }), _obj)), _obj)),

  Events: (_dec2 = (0, _decorators.deprecated)('Backendless.Events', 'Backendless.Events.dispatch'), (_obj2 = {
    dispatchSync: _utils2.default.synchronized(_dispatchEvent.dispatchEvent),
    dispatch: _utils2.default.promisified(_dispatchEvent.dispatchEvent)

  }, (_applyDecoratedDescriptor(_obj2, 'dispatchSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj2, 'dispatchSync'), _init2 = _init2 ? _init2.value : undefined, {
    enumerable: true,
    configurable: true,
    writable: true,
    initializer: function initializer() {
      return _init2;
    }
  }), _obj2)), _obj2))
};

exports.default = BL;
