// backendless/es/bl/invoke-service-method.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.invokeServiceMethod = invokeServiceMethod;

var _urls = require('../urls');

var _urls2 = _interopRequireDefault(_urls);

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function invokeServiceMethod(serviceName, method, parameters, executionType, asyncHandler) {
  if (typeof parameters === 'string') {
    asyncHandler = executionType;
    executionType = parameters;
    parameters = undefined;
  }

  if (executionType instanceof _async2.default) {
    asyncHandler = executionType;
    executionType = undefined;
  }

  if (parameters instanceof _async2.default) {
    asyncHandler = parameters;
    parameters = undefined;
  }

  var headers = {};

  if (executionType) {
    headers[_constants.EXECUTION_TYPE_HEADER] = executionType;
  }

  return _request2.default.post({
    url: _urls2.default.blServiceMethod(serviceName, method),
    data: parameters,
    isAsync: !!asyncHandler,
    headers: headers,
    asyncHandler: asyncHandler
  });
}
